redisinsight/api/src/modules/database-recommendation/providers/database-recommendation.provider.ts

```diff
@@ -0,0 +1,79 @@
+import { Injectable, Logger } from '@nestjs/common';
+import { DatabaseRecommendationEntity }
+  from 'src/modules/database-recommendation/entities/database-recommendation.entity';
+import { InjectRepository } from '@nestjs/typeorm';
+import { Repository } from 'typeorm';
+import { plainToClass } from 'class-transformer';
+import { RecommendationDto, RecommendationsDto } from 'src/modules/database-recommendation/dto';
+import { ClientMetadata } from 'src/common/models';
+
+@Injectable()
+export class DatabaseRecommendationProvider {
+  private readonly logger = new Logger('DatabaseRecommendationProvider');
+
+  constructor(
+    @InjectRepository(DatabaseRecommendationEntity)
+    private readonly repository: Repository<DatabaseRecommendationEntity>,
+  ) {}
+
+  /**
+   * Save entire entity
+   * @param databaseId
+   * @param recommendationName
+   */
+  async create(databaseId: string, recommendationName: string): Promise<RecommendationDto> {
+    this.logger.log('Creating database recommendation');
+    return await this.repository.save(
+      plainToClass(DatabaseRecommendationEntity, { databaseId, name: recommendationName }),
+    );
+  }
+
+  /**
+   * Return list of database recommendations
+   * @param clientMetadata
+   */
+  async list(clientMetadata: ClientMetadata): Promise<RecommendationsDto> {
+    this.logger.log('Getting database recommendations list');
+    const recommendations = await this.repository
+      .createQueryBuilder('r')
+      .where({ databaseId: clientMetadata.databaseId })
+      .select(['r.id', 'r.name'])
+      .orderBy('r.createdAt', 'DESC')
+      .getMany();
+
+    const totalUnread = await this.repository
+      .createQueryBuilder()
+      .where({ databaseId: clientMetadata.databaseId, read: false })
+      .getCount();
+
+    this.logger.log('Succeed to get recommendations');
+    return plainToClass(RecommendationsDto, {
+      recommendations,
+      totalUnread,
+    });
+  }
+
+  /**
+   * Return list of database recommendations
+   * @param databaseId
+   */
+  async read(databaseId: string): Promise<RecommendationsDto> {
+    this.logger.log('Getting database recommendations list');
+    await this.repository
+      .createQueryBuilder('r')
+      .update()
+      .where({ databaseId })
+      .set({ read: true })
+      .execute();
+
+    const totalUnread = await this.repository
+      .createQueryBuilder()
+      .where({ databaseId, read: false })
+      .getCount();
+
+    return new RecommendationsDto({
```

Comment:
not sure about response schema. Seems like will not match to **list** response if I'll call it right after **read**. What should happen after read all recommendations?

Reply:
But if we don't need to return notifications and totalUnread will be always 0 do we need to return anything then?